_posts/2019-02-19-refactoring-state-pattern.markdown

```diff
@@ -0,0 +1,114 @@
+---
+layout: post
+title:  "Refactoring with Design Patterns - The State Pattern"
+date: 2019-02-19 10:00:00
+categories: ["code-refactor", "design-patterns"]
+author: "cleiviane"
+---
+
+In this series of [code refactoring](https://www.ombulabs.com/blog/tags/code-refactor) posts we are discussing how design patterns can be used to make our Ruby code beautiful, maintainable and clean.
+
+Today I want to talk about a pattern that can be very useful when we need to control the flow of a set of events of our objects: *The State Pattern* or a.k.a *Finite State Machine*.
+
+As a developer it's something common to see objects changing of states. At the beginning manage the state of an object can be as simple as have some boolean attributes where you can check if the object is in state A or B. But when the complexity is increasing you can end up with a number of states that is difficulty to manage without break the [SOLID principles](http://butunclebob.com/ArticleS.UncleBob.PrinciplesOfOod). That is where we can benefit of the elegant solution the State Pattern give to us.
```

Comment:
"As a developer IT IS common to see objects changing states. At the beginning managING the state of an object can be as simple as havING some boolean attributes where you can check if the object is in state A or B. But when the complexity INCREASES you can end up with a number of states that ARE difficult to manage without breakING the SOLID principles. That is where we can benefit FROM the elegant solution the State Pattern giveS to us.